man/fi_tcp.7.md

```diff
@@ -32,6 +32,15 @@ The following features are supported
 *Shared Rx Context*
 : The tcp provider supports shared receive context
 
+# PROVIDER SPECIFIC OPERATION FLAGS
+: The tcp provider supports the following op flags
+
+*FI_TCP_NO_CONNECT*
+: This flag indicates that operations should fail if there is no
+  existing connection to the remote peer.  In such case, an FI_ENOTCONN
+  error should be expected.
```

Comment:
I would not make this a flag that's checked on every operation (i.e. it's okay to connect if it's a send, but not a write?).  It makes more sense either applied to the entire rdm endpoint or to a specific peer.

Reply:
in the target problematic scenario, the rdm is used for some eps where this flag is needed and some eps where it would not be, so applying to the entire rdm was undesirable.

Reply:
I think I also agree with the first part (ie. exposing low-level connection is a bad idea) though providing 2 types of endpoints is also difficult for the user. To provide some context as to what problem we were trying to solve, in a client-server configuration where the client is behind a firewall, we have cases where for example client A would send a message to server A, and server A would then attempt to do an emulated tcp RMA to client B through an fi_write() call (without prior connection established from client B to server A). In that case, it seems that with the tcp provider server A remains stuck attempting to establish a connection and doing an fi_cancel on the RMA does not seem to be able to complete, as it's not supported currently by the tcp provider. So what we wanted was a way of having some completion and error being returned when server A is not able to reach client B. I'm opened to other means but having to manage 2 endpoints seems also cumbersome, I would also be happy though if we don't have to expose any connection logic.

Reply:
Just to add a bit of context, this use case came from Parallelstore (Google's DAOS service).  We control only the server side of the equation and rely on the user to do client configuration (as we don't control their VMs, network configuration, and processes).  Opening the firewall to server to client connections is not common for services and requires them to do it explicitly or their writes will simply hang.  We want to remove this requirement as it is becoming a major scale issue for onboarding new customers.

Reply:
Client A is telling a server to issue an RMA write to client B?  A knows the memory information on client B, and B is expecting this transfer?  Is this the model used by the app?  Because if so, weird.

In any case, the entire RDM ep is behind a firewall, but connections can only go one direction?  If this is the case, then I think some generic EP setting might be able to convey this, such that the implementation can adjust.  The server could probably just fail the operation (asynchronously makes the most sense) in this situation.

I assume A is somehow notified what happened between the server and B?

Reply:
oh yes sorry I was trying to keep it simple and in the end that didn't make much sense, the real model is client A tells server A, which tells server B about its memory information, which then issues an RMA to client A.

Reply:
> In any case, the entire RDM ep is behind a firewall, but connections can only go one direction? If this is the case, then I think some generic EP setting might be able to convey this, such that the implementation can adjust. The server could probably just fail the operation (asynchronously makes the most sense) in this situation.

yeah I think adding a setting to the server EP would probably be fine in this case, although we'd still want servers to connect to each other using the same endpoint and not use a separate endpoint for intra-communication.

Reply:
I think we need to figure out the model for how firewalls should be handled, ignoring the implementation.  Maybe this is related to NAT as well.

An EP is assigned a single address.  Can traffic to that address go through the firewall?  Can traffic from that address go through the firewall?

Assume the implementation is using UDP (or some other connectionless transport).  Does that change anything about how the firewall must be configured?

If I'm understanding this now, what's being requested from the API is: Traffic is allowed to reach an EP (through a firewall), but traffic from that EP cannot reach the target ... unless it's already received some message from the target.  This assumes the EP is responsible for remembering which target addresses it has received data from, and that those addresses still reference the correct target process.

From the viewpoint of the API, this still seems weird.  I think the weirdness comes from the firewall acting on the lower-level connections, which the EP hides.

This is some per peer setting.  But if the application is responsible for providing the value of this setting, then it has an idea how to manage this.  In this example, server B could see if it knows A.  If not, fail sending to it.

One alternative is to configure the EP with firewall data only, and let the provider determine what to do with it.  In this case, a request to send to X checks the firewall config to see if it's possible.  Maybe X is another server within the firewall.  If so, connect and send.  If X is outside the firewall, fail the send.  I don't know what the firewall data would look like but providing it via some configuration file seems reasonable.  I.e. don't make this an application problem.

Reply:
That alternative is a bit is tricky.   The client may be older than the server and we want old clients to continue behaving as they do now (e..g if the user has already configured their firewall correctly and the client doesn't support getting back an error and retrying, we want it to just continue working).

Reply:
I think the problem is really on the client.  The client endpoint is the one behind the firewall and can't be reached (unless the server has already connected to it).   Can we encode something into the client side URI that would indicate it can't be reached?   The advantage to that approach from our end is that then the client could have complete control over telling the server whether or not it can handle the error (e.g. can support getting back an error indicating that the server couldn't connect and handle it appropriately)

Reply:
It's the server's behavior that should change.

From the viewpoint of the client, everything works.  It wants to talk to server A and can do so.  That server A wants to pass off the response to some other system that the client doesn't know about is related to the storage architecture, not the client SW.

I don't think pushing this detail into the apps is the best option.  But you can work-around this in the client by having the client send some sort of 'hello' message to every storage server during initialization -- to poke holes through the firewall.  That pushes the burden onto every client app that might want to use DAOS.

Alternatively, you can configure the server SW to be firewall aware, so that it avoids forwarding requests to servers not already communicating with the client.

Or, change the protocol around handling firewalls.  Have server A tell the client to retry its request with server B, rather than forwarding it internally.

There are likely other options for this.  But I would avoid picking one which encoded these details in the SW API.

Reply:
It would probably be better if this thread were copied into an issue for continued discussion, rather than attaching it to this PR.

Reply:
#10637